### PR TITLE
ARROW-16328: [Java] Proposal to upgrade Arrow Java project to JPMS Java Platform Module System

### PR DESCRIPTION
**Pull request by davisusanibar**
Proposal to upgrade Arrow Java to [Module System](https://openjdk.java.net/jeps/261).

Please note: This PR is not consider to be part of next release 10.0.0, the plan is to test this on 11.0.0 version.

**Current status**

1. Apache Arrow Java is compiled and tested by JDK 8, 11, 17, 18 but packaged with source/target JDK8

**Changes implemented by this PR**

1. Apache Arrow Java project are continuing compiled and tested by JDK 8, 11, 17, 18 but packaged with source/target JDK8
2. If Apache Arrow Java project is building with JDK11+ validate the project and test it against the module-info.java definitions
3. There are not CI implementation to deliver Jar library with JDK8 & JDK11+ (we need to define what will be our next step for JDK versions)
4. We probably are breaking APIs because we are changing package name to ensure unique package names on the project required by module-info.java
5. We are added a new project that contains Apache Arrow Netty custom implementation needed for
6. There are remaining modules pending to implement modules with module-info.java files. This PR contains modules-info.java for: Format / Memory / Vector
7. Is pending to review errors at Integration testing

**Next steps**

1. Implement cross-compilation with JDK 11+ and packaging with release=8 (continue offering support for JDK8):
That is not posible, if we compile with --release=8 we will have problems to access sun.misc.Unsafe that is not as exposed/public/documented API for N >= 9: -source N -target N --system <documented-APIs-from-N>.	(https://openjdk.org/jeps/247)
 
2. We have these options for our next steps:
- **Option 1**.- Maintain all as it is now, continue packaging with source/target JDK8

- **Option 2**.- Change/Update/Modify use of sun.misc.* classes dependencies to be able to support cross-compilation (--release 8) to packaging with JDK11+ and continue offering support for JDK8 also

````
jdeps --jdk-internals arrow-memory-unsafe-10.0.0-SNAPSHOT.jar
arrow-memory-unsafe-10.0.0-SNAPSHOT.jar -> jdk.unsupported
   org.apache.arrow.memory.unsafe.UnsafeAllocationManager -> sun.misc.Unsafe                                    JDK internal API (jdk.unsupported)

Warning: JDK internal APIs are unsupported and private to JDK implementation that are
subject to be removed or changed incompatibly and could break your application.
Please modify your code to eliminate dependence on any JDK internal APIs.
For the most recent update on JDK internal API replacements, please check:
https://wiki.openjdk.java.net/display/JDK8/Java+Dependency+Analysis+Tool

JDK Internal API                         Suggested Replacement
----------------                         ---------------------
sun.misc.Unsafe                          See http://openjdk.java.net/jeps/260

````

- **Option 3**.- _Drop support for JDK8_ and starting offering support to packaging with JDK11+, this will be the right approach, but, this thread does not have an agreement about to drop JDK8 support ([JAVA] JDK Support Policy - https://lists.apache.org/thread/phpgpydtt3yrgnncdyv4qdq1gf02s0yj)

If you decided to push some of this options please use that [ML](https://lists.apache.org/thread/phpgpydtt3yrgnncdyv4qdq1gf02s0yj) to move on next steps.

- - - - - - - - - - - - - - - - - - - - - - - 

**Detailed steps for current implementation**

JPMS Planning:

- [x] Migrate Arrow Java Format
- [x] Migrate Arrow Java Memory Core
- [x] Migrate Arrow Java Memory Netty
Not possible to apply module system for the package io.netty.buffer, it was needed to create another module `memory-netty-buffer-patch` and patch this package with our arrow java io.netty.buffer code.
- [x] Migrate Arrow Java Memory Unsafe
TODO: Need to refactor protected access methods from unit test.
- [x] Migrate Arrow Java Vector
TODO: Need to refactor protected access methods from unit test.
- [ ] Migrate Arrow Java Flight
Blocker: Current grpc-api and grpc-context are not modularized and cannot be used as a module because both of them has the same package named io.grpc

- **JPMS Java Platform Module System**
    
    Java modules enforced new restrictions on existing code:
    
    - Package need to be uniqueness
        - There are some modules with the same package name (i.e. org.apache.arrow.memory / io.netty.buffer) that is consumed by another module with the same package name to access **protected** methods.
            - Java Memory module problems:
                - Core / Netty / Unsafe has the same package name org.apache.arrow.memory and there are not problems to consume protected methods, if we change the name of the package we need to define a way about how Netty module could consume for example Core protected methods.
            - Java Netty extends io.netty.buffer capabilities:
                - Is needed to create a new module arrow-memory-netty-buffer-patch to add new Java functionalities added and this need to be consumed by Arrow Memory Netty.
                
    - Dependencies also need to be modularized:
        - Current jar **dependencies** need to be modularized to be consumed as a module if not there are no way to convert our project to module if any dependencies are not modularized before.
            - Arrow Format: [x]
            - Arrow Memory: [x]
            - Arrow Vector: [x]
            - Arrow Flight:
                - Blocker: Current grpc-api and grpc-context are not modularized and cannot be used as a module because both of them has the same package named io.grpc.
                
    
    Summarize changes and benefits to migrate Arrow Java with Module System:
    - JPMS offer Strong encapsulation, Well-defined interfaces, Explicit dependencies. (3)(4)
    - JPMS offer reliable configuration and security to hide platform internals.
    - JPMS offer a partial solution to solve problema about read (80%) /write (20%) code.
    - JPMS offer optimization for readability about read/write ratio (90/10) thru module-info.java.
    - Consistency logs, JPMS implement consistency logs to really use that to solve the current problem.
    - Be able to customize JRE needed with only modules needed (not java.desktop for example and others) thru JLink.
    - Modules is also been implemented by other language such as Javascript (ES2015), C++(C++20), Net (Nuget/NetCore) and this is some bus that at some time we could onboard.
    - Consider tho take a lokk at this discussion about pro/cons (5).

(3): https://nipafx.dev/java-modules-reflection-vs-encapsulation/
(4): https://github.com/nipafx/demo-jigsaw-reflection
(5): https://www.reddit.com/r/java/comments/okt3j3/do_you_use_jigsaw_modules_in_your_java_projects/

### CODE REVIEW

File: java/memory/memory-netty/pom.xml [hunk 1]
```diff
@@ -69,4 +70,25 @@
...
     </profile>
   </profiles>
+  <build>
+    <plugins>
+      <plugin>
+        <groupId>org.apache.maven.plugins</groupId>
+        <artifactId>maven-compiler-plugin</artifactId>
+        <configuration>
+          <compilerArgs combine.children="append">
+            <arg>--patch-module=io.netty.buffer=/Users/dsusanibar/.m2/repository/org/apache/arrow/memory-netty-buffer-patch/8.0.0-SNAPSHOT/memory-netty-buffer-patch-8.0.0-SNAPSHOT.jar</arg>
```
Comment (martin-g) [inline java/memory/memory-netty/pom.xml:80]: Use ` ${settings.localRepository}` instead of hardcoding the user

File: java/pom.xml [hunk 2]
```diff
@@ -28,18 +28,21 @@
...
-    <dep.netty.version>4.1.72.Final</dep.netty.version>
     <dep.jackson-bom.version>2.13.2.20220328</dep.jackson-bom.version>
+    <dep.netty.version>4.1.76.Final</dep.netty.version>
+    <dep.netty-bom.version>4.1.76.Final</dep.netty-bom.version>
     <dep.hadoop.version>2.7.1</dep.hadoop.version>
     <dep.fbs.version>1.12.0</dep.fbs.version>
     <dep.avro.version>1.10.0</dep.avro.version>
     <arrow.vector.classifier />
-    <forkCount>2</forkCount>
+    <forkCount>0</forkCount>
```
Comment (martin-g) [inline java/pom.xml:45]: Is this related to JPMS ?

File: java/memory/memory-netty/pom.xml (hunk 1 above)
Comment (davisusanibar) [reply java/memory/memory-netty/pom.xml]: Thanks, updated

File: java/pom.xml (hunk 2 above)
Comment (davisusanibar) [reply java/pom.xml]: To be reviewed. Value more than "0" produce: NoClassDefFound Could not initialize class org.apache.arrow.memory.RootAllocator

Comment (davisusanibar) [reply java/pom.xml]: Updated to 2

**Comment by jdimeo** [comment]
This looks like it's ready for review! Can we get it merged so that we can get Snowflake JDBC working with Java 17?
https://github.com/snowflakedb/snowflake-jdbc/issues/589

**Comment by json-warner** [comment]
> > This looks like it's ready for review! Can we get it merged so that we can get Snowflake JDBC working with Java 17? [snowflakedb/snowflake-jdbc#589](https://github.com/snowflakedb/snowflake-jdbc/issues/589)
> 
> Any update on this? Our team is also looking to have the Snowflake JDBC working with Java 17.

Same. It's been 5 mo since last update. Can someone provide a status? or timeline 🤞?

**Comment by amol-** [comment]
Closing because it has been untouched for a while, in case it's still relevant feel free to reopen and move it forward 👍

**Comment by jdimeo** [comment]
It's 100% still relevant. What's preventing the merge?

**Comment by danepitkin** [comment]
Arrow first needs to deprecate Java 8 support since Modules were introduced in Java 9. There's a recent discussion on this in the Arrow Mailing list[1].

[1] https://lists.apache.org/thread/s07jx58yw4mkl54t3bkggnyg0sftcrr8

**Comment by amoeba** [comment]
@danepitkin: In case it's applicable here, I recently noticed RoaringBitmap upgraded to JPMS w/o needing to drop support for Java 8, see https://github.com/RoaringBitmap/RoaringBitmap/pull/657. It looks like it requires dropping support for _building_ with Java 8 (must be built with Java 11+) but the result can still be run under Java 8. I'm not familiar with a lot of the details so I can't say if this same approach would help here but I thought I'd share.

**Comment by davisusanibar** [comment]
> @danepitkin: In case it's applicable here, I recently noticed RoaringBitmap upgraded to JPMS w/o needing to drop support for Java 8, see [RoaringBitmap/RoaringBitmap#657](https://github.com/RoaringBitmap/RoaringBitmap/pull/657). It looks like it requires dropping support for _building_ with Java 8 (must be built with Java 11+) but the result can still be run under Java 8. I'm not familiar with a lot of the details so I can't say if this same approach would help here but I thought I'd share.

Hi Team, yes, there are some solutions available for implementing JPMS ad target/release to JDK8, the main problem is that these solutions will fail if a project has Unsafe dependencies. Consider: https://github.com/apache/arrow/pull/13072#:~:text=Implement%20cross%2Dcompilation,jeps/247 and [JEP-247](https://openjdk.org/jeps/247#:~:text=For%20N%20%3E%3D%209%2C%20the,image%20is%20not%20allowed.).

The PR was moved around to include module-info.java files **to validate only JPMS requirements** and ensure that we are prepared to implement them after JDK8 deprecation.

**Comment by danepitkin** [comment]
Hey @mmendonca2 , I added an update to the issue you linked. In summary, the Arrow community decided not to deprecate Java 8 support for the foreseeable future. The plan is to revisit that discussion around 2024-06.

**Comment by jduo** [comment]
Hi all,

I'm investigating this now. It seems grpc recently got properly modularized in 1.57 (https://github.com/grpc/grpc-java/issues/3522) so we can rework the Flight modules.

As far as compiling the module-info.java files:
- There is a [Gradle plugin](https://github.com/Glavo/module-info-compiler) that can do a purely syntactic compilation of a module-info.java file into byte-code. It does not require rebuilding classes so I don't think we'll run into the Unsafe issues here. It will need to be rebuilt as a Maven plugin. This plugin does not handle annotations though, which does require classpath walking.
- A multi-module build using toolchans is suggested [here](https://stackoverflow.com/questions/53560584/java9-release-8-with-internal-packages-e-g-sun-misc-unsafe) for dealing with this problem as well. This might be more reliable since it uses official tooling, though it essentially doubles the compilation time (since you need to compile all source files twice).

**Comment by jduo** [comment]
I found this PR too difficult to rebase (there have been several changes to Netty and Memory classes and tests in 2023) so I started a new one based on this work at #38876)

**Comment by jduo** [comment]
This work has been migrated to #38876.

File: java/memory/memory-netty/pom.xml [hunk 3]
```diff
@@ -68,5 +75,25 @@
...
+        <jdk>[11,]</jdk>
+      </activation>
+            <build>
+              <plugins>
+                <plugin>
+                  <groupId>org.apache.maven.plugins</groupId>
+                  <artifactId>maven-compiler-plugin</artifactId>
+                  <configuration>
+                    <compilerArgs combine.children="append">
+                      <arg>--patch-module=io.netty.buffer=${project.basedir}/../memory-netty-buffer-patch/target/arrow-memory-netty-buffer-patch-${project.version}.jar</arg>
```
Comment (jduo) [inline java/memory/memory-netty/pom.xml:91]: @davisusanibar , I'm having problems getting this step to run in the new PR and am not entirely sure if it is necessary.

The new PR just does a pure syntax compilation of module-info.java so it doesn't verify the dependencies listed and just uses the classpath to look for classes from arrow-memory-netty-buffer-patch.

Was this step added so that this module would correctly find classes in io.netty.buffer defined in arrow-memory-netty-buffer-patch at build-time only, or does this have an effect at run time? My reading online is that even if we used patch-module to compile, a user utilizing the module would still need to use patch-module to specify module extensions to netty.

If patch-module is only used here to help get this JAR to compile, it seems like it isn't necessary in the new PR since the plugin is compiling module-info.java without verifying its imports.

**Comment by jayantsing-db** [comment]
> Hey @mmendonca2 , I added an update to the issue you linked. In summary, the Arrow community decided not to deprecate Java 8 support for the foreseeable future. The plan is to revisit that discussion around 2024-06.

Hi @danepitkin, it looks like the [PR](https://github.com/apache/arrow/issues/38051#issuecomment-2234222596) to drop Java 8 support has been merged. Could you update the status on [this comment](https://github.com/apache/arrow/pull/13072#issuecomment-1301079141) regarding Arrow's compatibility with JDK 16+? Currently, users still need to specify JVM parameters to open the required packages for reflection.

**Comment by lidavidm** [comment]
Unfortunately JVM parameters will be required for the foreseeable future. It may be possible to avoid this if someone can contribute support for FFM and work out how to support both Java >=21 and Java <21 from the same codebase.
